File: web/pythonScript/layout.py
from zone import Zone, Cartesian
from metazone import MetaZone, SuperCell
import copy
from typing import Optional, List, Tuple, Dict, Any


spec_header: Dict[str, Any] = {
    "name": "header",
    "ncells": 1,
    "nsteps": 1,
    "decpos": 1,
    "type": "decimal"
}


class Table1ArgZone_Arg(Zone):
    def __init__(self, spec: dict, info: Optional[dict] = None):
        super().__init__(info=info)
        self.spec = spec
        self.info = {
            'type': 1,
            'index': 0
        }
        self.zones: List[Tuple[SuperCell, Cartesian]]  # type: ignore
        for i in range(self.spec['args'][0]['nsteps']):
            positions = []
            for k in range(self.spec['args'][0]['ncells']):
                positions.append((1, k))
            self.add_zone(SuperCell(spec['args'][0], positions, info=self.info), (i, 0))

    def next_arg(self, arg: int, delta: int, index: int) -> SuperCell:
        return self.zones[index + delta][0]


class Table1ArgZone_Entry(Zone):
    def __init__(self, spec: dict, info: Optional[dict] = None):
        super().__init__(spec)
        self.spec = spec
        self.info = {
            'type': 0,
            'index': 0
        }
        self.zones: List[Tuple[SuperCell, Cartesian]]  # type: ignore
        for i in range(self.spec['args'][0]['nsteps']):
            positions = []
            for k in range(self.spec['entries'][0]['ncells']):
                positions.append((1, k))
            self.add_zone(SuperCell(spec['entries'][0], positions, info=self.info), (i, 0))

    def next_arg(self, arg: int, delta: int, index: int) -> SuperCell:
        return self.zones[index + delta][0]


class Table1ArgZone_Main(MetaZone):
    def __init__(self, spec: dict, info: Optional[dict] = None):
        super().__init__(info=info)
        self.spec = spec
        delta = 0
        if 'headers' not in self.info:
            self.info['headers'] = False
        if self.info['headers']:
            delta = 1
        self.add_zone(Table1ArgZone_Entry(spec, info=self.info), (delta, spec['args'][0]['ncells']))
        self.add_zone(Table1ArgZone_Arg(spec, info=self.info), (delta, 0))
        if self.info['headers']:
            self.add_zone(Table1ArgZone_HeaderTop(info={'length': self.C, 'name': self.spec['args'][0]['name'], 'entry_pos': self.spec['args'][0]['ncells']}), (0, 0))
        self.setup()

    def from_path(self, path: List[int]) -> Zone:
        if len(path) == 0:
            return self
        if path[0] == 0:
            return self.zones[0][0].from_path(path[2:])
        if path[0] == 1:
            return self.zones[1][0].from_path(path[2:])
        raise ValueError("Incorrect path coordinates")


class Table1ArgZone_HeaderTop(Zone):
    def __init__(self, info: dict):
        super().__init__(info=info)
        self.info = {
            'type': 2,
            'index': 0,
            'zone': 2
        }
        self.zones: List[Tuple[SuperCell, Cartesian]]  # type: ignore
        for i in range(info['length']):
            self.add_zone(SuperCell(spec_header, [(0, 0)], info=self.info), (0, i))
        self.zones[0][0].zones[0][0].val = info['name']
        self.zones[info['entry_pos']][0].zones[0][0].val = 'entry'

"""

"""


class Table2ArgZone_Arg0(Zone):
    def __init__(self, spec: dict, info: Optional[dict] = None):
        super().__init__(info=info)
        self.spec = spec
        self.info = {
            'type': 1,
            'index': 0,
            'zone': 0
        }
        self.zones: List[Tuple[SuperCell, Cartesian]]  # type: ignore
        for i in range(spec['args'][0]['nsteps']):
            positions = []
            for k in range(spec['args'][0]['ncells']):
                positions.append((0, k))
            self.add_zone(SuperCell(spec['args'][0], positions, info=self.info), (i, 0))

    def next_arg(self, arg: int, delta: int, index: int) -> Zone:
        if arg == 0:
            return self.zones[index + delta][0]
        else:
            return self

    def from_path(self, zone_coordinates: List[int]) -> Zone:
        if len(zone_coordinates) == 0:
            return self
        return self.zones[zone_coordinates[0]][0].from_path(zone_coordinates[2:])


class Table2ArgZone_Arg1(Zone):
    def __init__(self, spec: dict, info: Optional[dict] = None):
        super().__init__(info=info)
        self.spec = spec
        self.info = {
            'type': 1,
            'index': 1,
            'zone': 1
        }
        self.zones: List[Tuple[SuperCell, Cartesian]]  # type: ignore
        for i in range(spec['args'][1]['nsteps']):
            positions = []
            for k in range(spec['args'][1]['ncells']):
                positions.append((0, k))
            self.add_zone(SuperCell(spec['args'][1], positions, info=self.info), (0, i * spec['cwidth']))

    def next_arg(self, arg: int, delta: int, index: int) -> Zone:
        if arg == 1:
            return self.zones[index + delta][0]
        else:
            return self

    def from_path(self, zone_coordinates: List[int]) -> Zone:
        if len(zone_coordinates) == 0:
            return self
        return self.zones[zone_coordinates[1]][0].from_path(zone_coordinates[2:])


class Table2ArgZone_EntryLine(Zone):
    def __init__(self, spec: dict, info: Optional[dict] = None):
        super().__init__(info=info)
        self.spec = spec
        self.indexInParent: int
        self.parent: Zone
        for i in range(spec['args'][1]['nsteps']):
            positions = []
            for k in range(spec['entries'][0]['ncells']):
                positions.append((0, k))
            self.add_zone(SuperCell(spec['entries'][0], positions, info=self.info), (0, i * spec['cwidth']))

    def next_arg(self, arg: int, delta: int, index: int) -> Zone:
        if arg == 1:
            return self.zones[index + delta][0]
        else:
            if self.indexInParent + delta >= len(self.parent.zones):
                raise ValueError("Incorrect coordinates")
            return self.parent.zones[self.indexInParent + delta][0]


class Table2ArgZone_Entry(Zone):
    def __init__(self, spec: dict, info: Optional[dict] = None):
        super().__init__(info=info)
        self.spec = spec
        self.info = {
            'type': 0,
            'index': 0,
            'zone': 2
        }
        for i in range(spec['args'][0]['nsteps']):
            self.add_zone(Table2ArgZone_EntryLine(spec, info=self.info), (i, 0))


class Table2ArgZone_HeaderTop(Zone):
    def __init__(self, info: dict):
        super().__init__(info=info)
        self.info = {
            'type': 2,
            'index': 0,
            'zone': 2
        }
        self.zones: List[Tuple[SuperCell, Cartesian]]  # type: ignore
        spec = copy.deepcopy(spec_header)
        spec['ncells'] = info['cwidth']
        for i in range(info['number']):
            positions = []
            for c in range(spec['ncells']):
                positions.append((0, c))
            self.add_zone(SuperCell(spec_header, positions, info=self.info), (0, i * spec['ncells']))
        self.zones[info['number'] // 2][0].zones[0][0].val = info['name']


class Table2ArgZone_HeaderLeft(Zone):
    def __init__(self, info: dict):
        super().__init__(info=info)
        self.info = {
            'type': 2,
            'index': 0,
            'zone': 2
        }
        self.zones: List[Tuple[SuperCell, Cartesian]]  # type: ignore
        for i in range(info['length']):
            self.add_zone(SuperCell(spec_header, [(0, 0)], info=self.info), (i, 0))
        self.zones[self.R // 2][0].zones[0][0].val = info['name']


class Table2ArgZone_Main(MetaZone):
    def __init__(self, spec: dict, info: Optional[dict] = None):
        super().__init__(info=info)
        self.spec = spec
        self.spec['cwidth'] = max(spec['args'][1]['ncells'], spec['entries'][0]['ncells'])
        delta = 0
        if 'headers' not in self.info:
            self.info['headers'] = False
        if self.info['headers']:
            delta = 1
        self.add_zone(Table2ArgZone_Arg0(self.spec), (1 + delta, 0 + delta))
        self.add_zone(Table2ArgZone_Arg1(self.spec), (0 + delta, self.spec['args'][0]['ncells'] + delta))
        self.add_zone(Table2ArgZone_Entry(self.spec), (1 + delta, self.spec['args'][0]['ncells'] + delta))
        if self.info['headers']:
            self.add_zone(Table2ArgZone_HeaderTop({'number': self.spec['args'][1]['nsteps'], 'name': self.spec['args'][1]['name'], 'cwidth': self.spec['cwidth']}), (0, self.spec['args'][0]['ncells'] + 1))
            self.add_zone(Table2ArgZone_HeaderLeft({'length': self.R - 1, 'name': self.spec['args'][0]['name']}), (1, 0))
        self.setup()

    def from_path(self, path: List[int]) -> Zone:
        if len(path) == 0:
            return self
        if path[0] == 0:
            return self.zones[2][0].from_path(path[2:])
        if path[0] == 1 and path[1] == 0:
            return self.zones[0][0].from_path(path[2:])
        if path[0] == 1 and path[1] == 1:
            return self.zones[1][0].from_path(path[2:])
        return super().from_path(path)


if __name__ == '__main__':
    arg1_spec = {
        'args': [
            {
                'name': 'a',
                'ncells': 3,
                'nsteps': 90,
                'decpos': 1,
                'type': 'integer_and_sexagesimal'
            },
        ],
        'entries': [
            {
                'name': 'entry',
                'decpos': 1,
                'ncells': 4,
                'type': 'sexagesimal'
            }
        ],
        'cwidth': 4
    }

    main_zone = Table1ArgZone_Main(arg1_spec)

File: web/js/DISHAS_table/Layout.js
/* global Zone */
/* global CartesianCoordinates */
/* global SuperCell */
/* global MetaZone */


/* eslint-disable */
/**
 * Each \ :js:class:`Zone`\  containing \ :js:class:`SuperCell`\ s **MUST** implement such a method. It returns
 * the delta-th neighbour of a \ :js:class:`SuperCell`\  (specified by its index in the list of sub-\ :js:class:`Zone`\ s),
 * along the given argument.
 * @param  {Number} arg    index of the argument
 * @param  {Number} delta  the returned \ :js:class:`SuperCell`\  will be the delta-th neighbour
 * @param  {Number} index  index of the \ :js:class:`SuperCell`\ 
 * @return {SuperCell}     the delta-th neighbour
 */
function nextArg_example(arg, delta, index) {
}

if(false) {
    /**
     * A Layout \ :js:class:`MetaCell`\  takes as a parameter a specification of the table.
     * @example
     * {
     *     args: [
     *         {
     *             decpos: 1,
     *             name: "angle",
     *             ncells: 2,
     *             nsteps: 90,
     *             type: "sexagesimal"
     *         }
     *     ],
     *     entries: [
     *         {
     *             decpos: 1,
     *             name: "Entry",
     *             ncells: 4,
     *             type: "sexagesimal"
     *         }
     *     ]
     * }
     * @type {Object}
     */
    var spec_example = {
        args: [
            {
                decpos: 1,
                name: "angle",
                ncells: 2,
                nsteps: 90,
                type: "sexagesimal"
            }
        ],
        entries: [
            {
                decpos: 1,
                name: "Entry",
                ncells: 4,
                type: "sexagesimal"
            }
        ]
    };
}
/* eslint-enable */

/* ================================================== */
//implementation of the standard layout for 1 arg tables
class Table1ArgZone_Arg extends Zone {
    constructor(spec, info) {
        super(info);
        this.spec = spec;
        this.infos = {
            type: 1,
            index: 0
        };
        for(var i=0; i<spec.args[0].nsteps; i++) {
            var positions = [];
            for(var k=0; k<spec.args[0].ncells; k++) {
                positions.push(new CartesianCoordinates(0,k));
            }
            this.addZone(new SuperCell(spec.args[0], positions, {"argPos0" : i, "argPos" : [i]}), new CartesianCoordinates(i,0));
        }
    }
    nextArg(arg, delta, index) {
        return this.zones[index + delta];
    }
}

class Table1ArgZone_Entry extends Zone {
    constructor(spec, info) {
        super(info);
        this.spec = spec;
        this.infos = {
            type: 0,
            index: 0
        };
        for(var i=0; i<spec.args[0].nsteps; i++) {
            var positions = [];
            for(var k=0; k<spec.entries[0].ncells; k++) {
                positions.push(new CartesianCoordinates(0,k));
            }
            this.addZone(new SuperCell(spec.entries[0], positions, {"argPos0" : i, "argPos" : [i]}), new CartesianCoordinates(i,0));
        }
    }
    nextArg(arg, delta, index) {
        return this.zones[index + delta];
    }
}

class Table1ArgZone extends MetaZone {
    /**
     * RootZone for a standard layout 1-argument table
     * @param  {Object}  spec  Dictionary specifying the template of the table
     * @param  {Object}  info  Dictionary of additional custom information
     * @return {Table1ArgZone}
     */
    constructor(spec, info) {
        super(info, true);
        this.spec = spec;
        this.addZone(new Table1ArgZone_Entry(spec), new CartesianCoordinates(1,spec.args[0].ncells));
        this.addZone(new Table1ArgZone_Arg(spec), new CartesianCoordinates(1,0));
        this.setup();
        //this.buildGrid();
    }
    fromPath(path) {
        if(path.length === 0)
            return this;
        if(path[0] === 0)
            return this.zones[0].fromPath(path.slice(2));
        if(path[0] === 1)
            return this.zones[1].fromPath(path.slice(2));
    }
}

/* ================================================== */
//implementation of the standard layout for 2arg tables
class Table2ArgZone_Arg0 extends Zone {
    constructor(spec, info) {
        super(info);
        this.spec = spec;
        this.infos = {
            type: 1,
            index: 0,
            zone: 0
        };
        for(var i=0; i<spec.args[0].nsteps; i++) {
            var positions = [];
            for(var k=0; k<spec.args[0].ncells; k++) {
                positions.push(new CartesianCoordinates(0,k));
            }
            this.addZone(new SuperCell(spec.args[0], positions, {"argPos0" : i, "argPos" : [i, 0]}), new CartesianCoordinates(i,0));
        }
    }
    nextArg(arg, delta, index) {
        if(arg === 0)
            return this.zones[index + delta];
        else {
            return this;
        }
    }
    fromPath(path) {
        if(path.length === 0)
            return this;
        return this.zones[path[0]].fromPath(path.slice(2));
    }
}

class Table2ArgZone_Arg1 extends Zone {
    constructor(spec, info) {
        super(info);
        this.spec = spec;
        this.infos = {
            type: 1,
            index: 1,
            zone: 1
        };
        for(var i=0; i<spec.args[1].nsteps; i++) {
            var positions = [];
            for(var k=0; k<spec.args[1].ncells; k++) {
                positions.push(new CartesianCoordinates(0,k));
            }
            this.addZone(new SuperCell(spec.args[1], positions, {"argPos1" : i, "argPos" : [0, i]}), new CartesianCoordinates(0,i*spec.cwidth));
        }
    }
    nextArg(arg, delta, index) {
        if(arg === 1)
            return this.zones[index + delta];
        else {
            return this;
        }
    }
    fromPath(path) {
        if(path.length === 0)
            return this;
        return this.zones[path[1]].fromPath(path.slice(2));
    }
}

class Table2ArgZone_Entry extends Zone {
    constructor(spec, info) {
        super(info);
        this.spec = spec;
        this.infos = {
            type: 0,
            index: 0,
            zone: 2
        };
        var that = this;
        for(var i=0; i<spec.args[0].nsteps; i++) {
            this.addZone(new Table2ArgZone_EntryLine(spec, {"argPos0" : i}), new CartesianCoordinates(i,0));
        }
        /*forWrapper(spec.args[0].nsteps, 0, 100,
            function(i) {
                that.addZone(new Table2ArgZone_EntryLine(spec, {argPos0 : i}), new CartesianCoordinates(i,0));
            }, function() {
            }, 5);*/
    }
}

class Table2ArgZone_EntryLine extends Zone {
    constructor(spec, info) {
        super(info);
        this.spec = spec;
        for(var i=0; i<spec.args[1].nsteps; i++) {
            var positions = [];
            for(var k=0; k<spec.entries[0].ncells; k++) {
                positions.push(new CartesianCoordinates(0,k));
            }
            this.addZone(new SuperCell(spec.entries[0], positions, {"argPos1" : i, "argPos" : [this.infos.argPos0, i]}), new CartesianCoordinates(0,i*spec.cwidth));
        }
    }
    nextArg(arg, delta, index) {
        if(arg === 1)
            return this.zones[index + delta];
        else {
            if(this.parent.zones[this.indexInParent + delta] === undefined)
                return null;
            return this.parent.zones[this.indexInParent + delta].zones[index];
        }
    }
}

class Table2ArgZone extends MetaZone {
    /**
     * RootZone for a standard layout 2-arguments table
     * @param  {Object}  spec  Dictionnary specifying the template of the table
     * @param  {Object}  info  Dictionnary of additionnal custom informations
     * @return {Table2ArgZone}
     */
    constructor(spec, info) {
        super(info, true);
        this.spec = spec;
        this.spec.cwidth = Math.max(spec.args[1].ncells, spec.entries[0].ncells);
        this.addZone(new Table2ArgZone_Entry(this.spec), new CartesianCoordinates(1,this.spec.args[0].ncells));
        this.addZone(new Table2ArgZone_Arg0(this.spec), new CartesianCoordinates(1,0));
        this.addZone(new Table2ArgZone_Arg1(this.spec), new CartesianCoordinates(0,this.spec.args[0].ncells));
        this.setup();
        //this.buildGrid();
    }
    fromPath(path) {
        if(path.length === 0)
            return this;
        if(path[0] === 0)
            return this.zones[0].fromPath(path.slice(2));
        if(path[0] === 1 && path[1] === 0)
            return this.zones[1].fromPath(path.slice(2));
        if(path[0] === 1 && path[1] === 1)
            return this.zones[2].fromPath(path.slice(2));
    }
}

/* =================================== */
// Definition of virtual zones for computing 2-args quantities
// spec will be the same as the table, with enties changed (with ncells = 1; type="decimal")

class ComputationZone2Arg extends MetaZone {
    constructor(spec, info) {
        super(info, true);
        this.spec = spec;

        this.addZone(new Table2ArgZone_Entry(this.spec), new CartesianCoordinates(0,0));
        
        this.setup();
    }
    fromPath(path) {
        if(path.length === 0)
            return this;
        if(path[0] === 0)
            return this.zones[0].fromPath(path.slice(2));
    }
}

class ComputationZone1Arg extends MetaZone {
    constructor(spec, info) {
        super(info, true);
        this.spec = spec;
        
        this.addZone(new Table1ArgZone_Entry(this.spec), new CartesianCoordinates(0,0));
        
        this.setup();
    }
    fromPath(path) {
        if(path.length === 0)
            return this;
        if(path[0] === 0)
            return this.zones[0].fromPath(path.slice(2));
    }
}

/* =================================== */
// Definition of graphical info zones

class VerticalInfoZone extends MetaZone {
    constructor(spec, info) {
        if(info === undefined)
            info = {nargs : 2};
        //info.type = 2;
        info.index = 0;
        super(info, true);

        this.spec = spec;
        spec.entries[0].nsteps = spec.args[0].nsteps;

        var zoneInfo = {};
        for(var key in info) {
            zoneInfo[key] = info[key];
        }
        zoneInfo.type = 2;
        this.addZone(new VerticalInfoZone_Entry(spec.entries[0], zoneInfo), new CartesianCoordinates(0,0));
        if(info.show_args) {
            var positions = [];
            for(var k=0; k<spec.args[1].ncells; k++) {
                positions.push(new CartesianCoordinates(0,k));
            }
            this.addZone(new SuperCell(spec.args[1], positions, {"type": 3, "argPos": 0}), new CartesianCoordinates(0,0));
        }

        this.setup();
    }
}

class VerticalInfoZone_Entry extends Zone {
    constructor(spec, info) {
        if(info === undefined)
            info = {nargs : 2};
        info.type = 2;
        info.index = 0;
        super(info, true);
        this.spec = spec;
        
        for(var i=0; i<spec.nsteps; i++) {
            var positions = [];
            for(var k=0; k<spec.ncells; k++) {
                positions.push(new CartesianCoordinates(0,k));
            }
            this.addZone(new SuperCell(spec, positions, {"argPos0" : i}), new CartesianCoordinates(i+((this.infos.nargs === 1)?1:1),0));
        }
    }
}

class HorizontalInfoZone extends MetaZone {
    constructor(spec, info) {
        if(info === undefined)
            info = {};
        //info.type = 2;
        info.index = 0;
        super(info, true);
        this.spec = spec;
        
        var zoneInfo = {};
        for(var key in info) {
            zoneInfo[key] = info[key];
        }
        zoneInfo.type = 2;
        this.addZone(new HorizontalInfoZone_Entry(spec, zoneInfo), new CartesianCoordinates(0,0));
        if(info.show_args) {
            var positions = [];
            for(var k=0; k<spec.args[0].ncells; k++) {
                positions.push(new CartesianCoordinates(0,k));
            }
            this.addZone(new SuperCell(spec.args[0], positions, {"type": 1, "argPos": 0}), new CartesianCoordinates(0,0));
        }
        this.setup();
    }
}

class HorizontalInfoZone_Entry extends Zone {
    constructor(spec, info) {
        if(info === undefined)
            info = {};
        info.index = 0;
        super(info, true);
        this.spec = spec;
        var delta = 1;
        if(info.show_args) {
            delta = 0;
        }
        for(var i=0; i<spec.args[1].nsteps; i++) {
            var positions = [];
            for(var k=0; k<spec.entries[0].ncells; k++) {
                positions.push(new CartesianCoordinates(0,k));
            }
            this.addZone(new SuperCell(spec.entries[0], positions, {"argPos1" : i}), new CartesianCoordinates(0,i*spec.entries[0].ncells+spec.args[0].ncells + delta));
        }
    }
}


///////////////////////////////////

/* ================================================== */
//implementation of an alternative layout for 1arg tables
class Alternative_Table1ArgZone_Arg extends Zone {
    constructor(spec, info) {
        super(info);
        this.spec = spec;
        this.infos = {
            type: 1,
            index: 0
        };
        for(var i=0; i<spec.args[0].nsteps; i++) {
            var positions = [];
            for(var k=0; k<spec.args[0].ncells; k++) {
                positions.push(new CartesianCoordinates(0,k));
            }
            this.addZone(new SuperCell(spec.args[0], positions, {argPos0 : i}), new CartesianCoordinates(0,i*spec.args[0].ncells));
        }
    }
    nextArg(arg, delta, index) {
        return this.zones[index + delta];
    }
}

class Alternative_Table1ArgZone_Entry extends Zone {
    constructor(spec, info) {
        super(info);
        this.spec = spec;
        this.infos = {
            type: 0,
            index: 0
        };
        for(var i=0; i<spec.args[0].nsteps; i++) {
            var positions = [];
            for(var k=0; k<spec.entries[0].ncells; k++) {
                positions.push(new CartesianCoordinates(0,k));
            }
            this.addZone(new SuperCell(spec.entries[0], positions, {"argPos0" : i}), new CartesianCoordinates(0,i*spec.args[0].ncells));
        }
    }
    nextArg(arg, delta, index) {
        return this.zones[index + delta];
    }
}

class Alternative_Table1ArgZone extends MetaZone {
    constructor(spec, info) {
        super(info, true);
        this.spec = spec;
        this.addZone(new Alternative_Table1ArgZone_Entry(spec), new CartesianCoordinates(1,0));
        this.addZone(new Alternative_Table1ArgZone_Arg(spec), new CartesianCoordinates(0,0));
        this.setup();
        //this.buildGrid();
    }
    fromPath(path) {
        if(path.length === 0)
            return this;
        if(path[0] === 0)
            return this.zones[0].fromPath(path.slice(2));
        if(path[0] === 1)
            return this.zones[1].fromPath(path.slice(2));
    }
}
